### ISSUE TITLE
exit while loop in examples/test_websocket.py

### ISSUE THREAD

**Issue by grantiago** [bachya/aioambient#131]
Hi. Sorry to use this for Q&A. I have no experience with websockets.  The ambient weather REST api fails more often than not. I took look at this test_websocket.py file and I can save the json results as needed. My issue is, I have been at this for a while now,  I haven't been able to find a way to exit from the while True: loop on line 62.  I need to exit the program. Thanks for any help and understanding.

**Comment by bachya** [comment]
Have you looked at [the example in the README]https://github.com/bachya/aioambient#websocket-api) (different from `examples/test_websocket.py`)? That shows you how to connect to the websocket, then disconnect from it later.

**Comment by grantiago** [comment]
Thanks for the reply. I'm working on it. I had to get python 3.9 up and running in a venv as you suggested. I thought my errors might be coming from trying to run it on 3.8. 

3.8 errors on File https://github.com/bachya/aioambient#websocket-api) 
```
./wSocket.py", line 45, in main
    websocket.on_disconnect(disconnect_method.data)
AttributeError: 'function' object has no attribute 'data'
```
I now have 3.9 running in a cloned venv:
when I ran the 'script/setup' 
`Failed to create the collection: Prompt dismissed..`
from the websocket example copied and edited from readme

```
 File "/home/grant/dev/weather/venv9/lib/python3.9/site-packages/socketio/asyncio_client.py", line 445, 
   in _trigger_event
    ret = handler(*args)
TypeError: disconnect_method() missing 1 required positional argument: 'data'
(venv9) grant@hillpi:~/dev/weather/venv9/aioambient$
```

thanks for your time.

**Comment by bachya** [comment]
Please share your entire script; if I can see more detail, I may be able to help more effectively.

**Comment by grantiago** [comment]
Hi. ~~Well after all this I have the test_rest_api.py working. Which answers my needs.~~  
~~Maybe the API was having problems over the holidays. Or maybe it was the phase of the moon.~~

**Update** Well that was an hour of glory. The rest API is failing again. It does seem to respond intermittently. Maybe my best option is to run the rest script every hour and hope ambient weather responds a couple times a day. It's not as though this is critical, but it is nice to have a degree of reliability. 

```ERROR:root:There was an error: Error requesting data from https://rt.ambientweather.net/v1/devices: 524, message='' . . .```

If you could help with the websocket I will continue with that. It looks as if for any degree of reliability The websocket is going to be the answer.  

I cloned your repository and pushed two scripts into the examples directory:

[examples](https://github.com/grantiago/aioambient/tree/dev/examples)

[readme_socket.py](https://github.com/grantiago/aioambient/blob/dev/examples/readme_socket.py)

[websocketx.py](https://github.com/grantiago/aioambient/blob/dev/examples/websocketx.py)

I'm a "hello world" kind of python coder. My scripts are edited copies of what I found here in the repo. 
**readme_socket.py**  is copied from the readme.md exits w/ the error `TypeError: disconnect_method() missing 1 required positional argument: 'data'
`
**websocketx.py** is a copy of examples/test_websocket.py which runs and writes the json I need. But it just keeps running ...

Thanks for your time. 

 *keys removed*

**Comment by bachya** [comment]
I don't get that `TypeError` in either the example or the readme snippet. Have you set up the dev environment per the instructions at the bottom of the readme?

**Comment by grantiago** [comment]
Thanks. I'm sorry I thought I had mentioned that.  Running `script/setup` results in: 

```(venv9) grant@hillpi:~/dev/weather/venv9/aioambient$ script/setup
Requirement already satisfied: poetry in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (1.3.1)
Requirement already satisfied: urllib3<2.0.0,>=1.26.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (1.26.13)
Requirement already satisfied: crashtest<0.5.0,>=0.4.1 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (0.4.1)
Requirement already satisfied: poetry-plugin-export<2.0.0,>=1.2.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (1.2.0)
Requirement already satisfied: requests<3.0,>=2.18 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (2.28.1)
Requirement already satisfied: virtualenv!=20.4.5,!=20.4.6,<21.0.0,>=20.4.3 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (20.17.1)
Requirement already satisfied: cachecontrol[filecache]<0.13.0,>=0.12.9 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (0.12.11)
Requirement already satisfied: packaging>=20.4 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (22.0)
Requirement already satisfied: keyring<24.0.0,>=23.9.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (23.13.1)
Requirement already satisfied: importlib-metadata<5.0,>=4.4 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (4.13.0)
Requirement already satisfied: tomli<3.0.0,>=2.0.1 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (2.0.1)
Requirement already satisfied: jsonschema<5.0.0,>=4.10.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (4.17.3)
Requirement already satisfied: trove-classifiers>=2022.5.19 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (2022.12.22)
Requirement already satisfied: lockfile<0.13.0,>=0.12.2 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (0.12.2)
Requirement already satisfied: filelock<4.0.0,>=3.8.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (3.9.0)
Requirement already satisfied: poetry-core==1.4.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (1.4.0)
Requirement already satisfied: tomlkit!=0.11.2,!=0.11.3,<1.0.0,>=0.11.1 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (0.11.6)
Requirement already satisfied: dulwich<0.21.0,>=0.20.46 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (0.20.50)
Requirement already satisfied: pexpect<5.0.0,>=4.7.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (4.8.0)
Requirement already satisfied: requests-toolbelt<0.11.0,>=0.9.1 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (0.10.1)
Requirement already satisfied: shellingham<2.0,>=1.5 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (1.5.0)
Requirement already satisfied: cleo<3.0.0,>=2.0.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (2.0.1)
Requirement already satisfied: pkginfo<2.0,>=1.5 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (1.9.2)
Requirement already satisfied: html5lib<2.0,>=1.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (1.1)
Requirement already satisfied: platformdirs<3.0.0,>=2.5.2 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from poetry) (2.6.2)
Requirement already satisfied: msgpack>=0.5.2 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from cachecontrol[filecache]<0.13.0,>=0.12.9->poetry) (1.0.4)
Requirement already satisfied: rapidfuzz<3.0.0,>=2.2.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from cleo<3.0.0,>=2.0.0->poetry) (2.13.7)
Requirement already satisfied: six>=1.9 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from html5lib<2.0,>=1.0->poetry) (1.16.0)
Requirement already satisfied: webencodings in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from html5lib<2.0,>=1.0->poetry) (0.5.1)
Requirement already satisfied: zipp>=0.5 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from importlib-metadata<5.0,>=4.4->poetry) (3.11.0)
Requirement already satisfied: attrs>=17.4.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from jsonschema<5.0.0,>=4.10.0->poetry) (22.2.0)
Requirement already satisfied: pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from jsonschema<5.0.0,>=4.10.0->poetry) (0.19.3)
Requirement already satisfied: jeepney>=0.4.2 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from keyring<24.0.0,>=23.9.0->poetry) (0.8.0)
Requirement already satisfied: jaraco.classes in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from keyring<24.0.0,>=23.9.0->poetry) (3.2.3)
Requirement already satisfied: SecretStorage>=3.2 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from keyring<24.0.0,>=23.9.0->poetry) (3.3.3)
Requirement already satisfied: ptyprocess>=0.5 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from pexpect<5.0.0,>=4.7.0->poetry) (0.7.0)
Requirement already satisfied: charset-normalizer<3,>=2 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from requests<3.0,>=2.18->poetry) (2.1.1)
Requirement already satisfied: certifi>=2017.4.17 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from requests<3.0,>=2.18->poetry) (2022.12.7)
Requirement already satisfied: idna<4,>=2.5 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from requests<3.0,>=2.18->poetry) (3.4)
Requirement already satisfied: distlib<1,>=0.3.6 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from virtualenv!=20.4.5,!=20.4.6,<21.0.0,>=20.4.3->poetry) (0.3.6)
Requirement already satisfied: cryptography>=2.0 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from SecretStorage>=3.2->keyring<24.0.0,>=23.9.0->poetry) (39.0.0)
Requirement already satisfied: more-itertools in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from jaraco.classes->keyring<24.0.0,>=23.9.0->poetry) (9.0.0)
Requirement already satisfied: cffi>=1.12 in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from cryptography>=2.0->SecretStorage>=3.2->keyring<24.0.0,>=23.9.0->poetry) (1.15.1)
Requirement already satisfied: pycparser in /home/grant/dev/weather/venv9/lib/python3.9/site-packages (from cffi>=1.12->cryptography>=2.0->SecretStorage>=3.2->keyring<24.0.0,>=23.9.0->poetry) (2.21)
Updating dependencies
Resolving dependencies... (0.5s)

Failed to create the collection: Prompt dismissed..```

On the bright side, I'm running the rest api now with some success. 

Is the api going out and retrieving up the details from the remote weather station? Or is it stored on an AmbientWeather server? 

One of my weather stations is on a mountainside in Mexico using  a slow satellite connection. I noticed as I loop through the responses the station here in the US writes to the file almost instantly. The second device, in Mexico, has a long pause before it writes. It varies. The last call was almost instant for both stations.

**Comment by bachya** [comment]
I've never seen `Failed to create the collection: Prompt dismissed..` before—what OS are you using?

> Is the api going out and retrieving up the details from the remote weather station? Or is it stored on an AmbientWeather server?

This library talks to the Ambient Weather cloud API—so, the device sends data at regular intervals to that API, then this library queries the API.

**Comment by grantiago** [comment]
I'm on a headless raspberry pi 4 running ubuntu 20.04 focal server
```lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.5 LTS
Release:        20.04
Codename:       focal
```
```
whereis python
python: 
/usr/bin/python3.8 
/usr/bin/python3.8-config 
/usr/bin/python3.9
/usr/lib/python2.7 
/usr/lib/python3.8 
/usr/lib/python3.9 
/etc/python2.7 
/etc/python3.8 
/etc/python3.9 
/usr/local/lib/python3.8 
/usr/local/lib/python3.9 
/usr/include/python3.8

**Comment by bachya** [comment]
Hmm, okay. Not quite sure what's going on with your environment, but let's focus on what is working.

> **websocketx.py** is a copy of examples/test_websocket.py which runs and writes the json I need. But it just keeps running ...

As you point out, you need to break out of the while loop—as to when that should happen, that's up to you and your program. Are you wanting to measure a single measurement, then quit? You could keep a counter variable, increment it inside the callback provided to `on_data`, and then, in the loop, `break` when that counter reaches a certain value.

**Comment by grantiago** [comment]
Hi. Thanks again. I made some changes on my test server. I added a swap file and turned off all the logger output. That seems to have solved the errors with the rest API. I think it was running out of memory (though it has 8gb ram) I noticed an increase in speed after turning off the logger. A quick search on SO found a couple posts saying to add the swap file on Ubuntu servers. 
However, it didn't get rid of the environment problems. I tried a counter on the websocket earlier and couldn't get it to work. I'll have another go at it later.
